File: LeetCode/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] cands, int target) {
        List<List<Integer>> res = new ArrayList<>();
        class Helper {
            void backtrack(int i, List<Integer> cur, int diff) {
                if (diff == 0) res.add(new ArrayList<>(cur));
                if (i >= cands.length || diff <= 0) return;
                cur.add(cands[i]);
                backtrack(i, cur, diff - cands[i]);
                cur.remove(cur.size() - 1);
                backtrack(i + 1, cur, diff);
            }
        }
        new Helper().backtrack(0, new ArrayList<>(), target);
        return res;
    }
} //TC: O(N^(M/T + 1)), SC: O(T/M), where T is the target value, N is the number of candidates, M is the smallest candidate among all the given integers

File: LeetCode/39. Combination Sum.kt
class Solution {
    fun combinationSum(cs: IntArray, target: Int): List<List<Int>> {
        val res = mutableListOf<List<Int>>()

        fun backtrack(remain: Int,
                      i: Int = 0,
                      tmp: MutableList<Int> = mutableListOf()) {
            if (remain < 0 || i >= cs.size) return //sum is over target value or iterate is going over candidates
            if (remain == 0) res.add(tmp.toList()) //combination reaches target value, add it to result, toList() to make a copy of current temp list
            else {
                tmp.add(cs[i]) //add current current iterate of candidates to temp list
                backtrack(remain - cs[i], i, tmp) //backtrack with remaining number, with current iterate because we can reuse the same number, going deeper in a decision tree with this method
                tmp.removeAt(tmp.size - 1) //restore to previous state for another decision after backtracking, as in java, list is passed by reference
                backtrack(remain, i+1, tmp) //use the next candidate for backtrack, going broader in a decision tree with this method
            }
        }

        backtrack(target) //initial target value
        return res
    }
}
